observer/baseNodeProvider.go

```diff
@@ -28,26 +29,173 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
 		newNodes[shardId] = append(newNodes[shardId], observer)
 	}
 
-	bop.mutNodes.Lock()
-	bop.nodes = newNodes
-	bop.allNodes = initAllNodesSlice(newNodes)
-	bop.mutNodes.Unlock()
+	bnp.mutNodes.Lock()
+	bnp.nodes = newNodes
+	bnp.allNodes = initAllNodesSlice(newNodes)
```

Comment:
because for example there are a lot of endpoints that need the observers from shard X. if we didn't have this map, we'd have to compute a list each time (by iterating over all the observers and selecting the ones in the desired shard)

Reply:
ok. here i was thinking to have maybe only the map, and get the slice with all nodes only when needed (if they are truly in sync), this way we don't have to remove/add for both, or maybe iterating by shards every it's fine; maybe in a different PR if it's relevant